lldb/test/Shell/helper/stepper.py

```diff
@@ -0,0 +1,213 @@
+
+# Skip doing "po" on all locals, in each frame.
+SKIP_PO = False
+
+# Maximum number of times to visit a PC before "finish"ing out.
+MAX_VISITS_PER_PC = 50
+
+# Run the process. It's useful to set this to False if you have monitor/test
+# lldb's, and the child lldb can't read from stdin to do things like continue
+# at a breakpoint.
+RUN_THE_PROCESS = True
+
+# Skip N frames between each inspection.
+SKIP_N_FRAMES_BETWEEN_INSPECTIONS = 0
+
+###############################################################################
+
+
+def parse_options():
+    """
+    Parse the STEPPER_OPTIONS environment variable.
+
+    Update any constants specified in the option string.
+    """
+    global ONLY_VISIT_UNOPTIMIZED, ONLY_VISIT_SWIFT, SKIP_FRAMEVAR, \
+            SKIP_PO, MAX_VISITS_PER_PC, RUN_THE_PROCESS, \
+            SKIP_N_FRAMES_BETWEEN_INSPECTIONS
+    opts = os.getenv('STEPPER_OPTIONS', '').split(';')
+    for o in opts:
+        m = re.match('(\w+)=([^;]+)', o)
+        if not m:
+            print('Unrecognized option:', o)
+        else:
+            option, val = m.groups()
+            print('Setting', option, '=', val)
+            globals()[option] = eval(val)
+
+
+def doit(dbg, cmd):
+    "Run a driver command."
+    print('::', cmd)
+    dbg.HandleCommand(cmd)
+    return False
+
+
+def doquit(dbg):
+    "Quit the stepper script interpreter."
+    os._exit(0)
+
+
+def should_stop_stepping(process):
+    "Decide whether we should stop stepping."
+    state = process.GetState()
+    if state in (lldb.eStateExited, lldb.eStateDetached):
+        print('Process has exited or has been detached, exiting...')
+        return True
+    if state in (lldb.eStateCrashed, lldb.eStateInvalid):
+        print('Process has crashed or is in an invalid state, exiting...')
+        return True
+    return False
+
+
+def alter_PC(dbg, process, cmd):
+    "Run a driver command that changes the PC. Return whether to stop stepping."
+    # Check the process state /after/ we step. Any time we advance the PC,
+    # the process state may change.
+    doit(dbg, cmd)
+    return should_stop_stepping(process)
+
+
+def return_from_frame(thread, frame):
+    print(':: Popping current frame...')
+    old_name = frame.GetFunctionName()
+    thread.StepOutOfFrame(frame)
+    new_frame = thread.GetSelectedFrame()
+    new_name = new_frame.GetFunctionName()
+    print(':: Transitioned from {0} -> {1}.'.format(old_name, new_name))
+    return True
+
+
+def __lldb_init_module(dbg, internal_dict):
+    parse_options()
+
+    # Make each debugger command synchronous.
+    dbg.SetAsync(False)
+
+    # Run the program and stop it when it reaches main().
+    if RUN_THE_PROCESS:
+        doit(dbg, 'breakpoint set -n main')
+        doit(dbg, 'run')
+
+    # Step through the program until it exits.
+    gen = 0
+    target = dbg.GetSelectedTarget()
+    process = target.GetProcess()
+    visited_pc_counts = defaultdict(int)
+    while True:
+        gen += 1
```

Comment:
would it be useful to also track how many times `do_inspection` ends up being true, and including that count along with the generation count?

Reply:
Yep, we want to configure tests to minimize the difference between #inspections and #generations, otherwise we're just wasting cycles without gaining test coverage.